File: spoj/BRCKTS.py
class Node:
	open_bracket_count = 0
	closed_bracket_count = 0
	total_sum = 0
	start_index = 0
	end_index = 0  

	def __init__(self):
		pass

def create_tree(tree, input_string, current_node_index, start_index, end_index):

	if start_index > end_index:
		return

	if start_index == end_index:
		new_node = Node()

		if input_string[start_index] == '(':
			new_node.open_bracket_count = 1
			new_node.closed_bracket_count = 0
			new_node.total_sum = 1
		else:
			new_node.open_bracket_count = 0
			new_node.closed_bracket_count = 1
			new_node.total_sum = -1

		new_node.start_index = start_index
		new_node.end_index = start_index

		tree[current_node_index] = new_node

		return
	
	mid = (start_index + end_index) / 2
	create_tree(tree, input_string, current_node_index*2 + 1, start_index, mid)
	create_tree(tree, input_string, current_node_index*2 + 2, mid + 1, end_index)

	left_child =  tree[current_node_index*2 + 1];
	right_child = tree[current_node_index*2 + 2];

	curr_node = Node()
	curr_node.open_bracket_count = left_child.open_bracket_count + right_child.open_bracket_count;
	curr_node.closed_bracket_count = left_child.closed_bracket_count + right_child.closed_bracket_count;
	curr_node.total_sum = left_child.total_sum + right_child.total_sum;
	curr_node.start_index = left_child.start_index;
	curr_node.end_index = right_child.end_index;

	tree[current_node_index] = curr_node;


def update_tree(tree, update_index):
	if update_index < 0:
		return

	curr_node = tree[update_index]

	if curr_node.start_index == curr_node.end_index:
		if curr_node.open_bracket_count == 1:
			curr_node.open_bracket_count = 0
			curr_node.total_sum = -1
			curr_node.closed_bracket_count = 1
		else:
			curr_node.open_bracket_count = 1
			curr_node.total_sum = 1
			curr_node.closed_bracket_count = 0


	else:
		left_child =  tree[update_index*2 + 1];
		right_child = tree[update_index*2 + 2];

		curr_node.open_bracket_count = left_child.open_bracket_count + right_child.open_bracket_count;
		curr_node.closed_bracket_count = left_child.closed_bracket_count + right_child.closed_bracket_count;
		curr_node.total_sum = left_child.total_sum + right_child.total_sum;
		curr_node.start_index = left_child.start_index;
		curr_node.end_index = right_child.end_index;

	tree[update_index] = curr_node

	if update_index % 2 == 0:
		update_tree(tree, update_index/2 -1)
	else:
		update_tree(tree, update_index/2)

for count in range (0, 10):
	length = int(raw_input())
	input_string = raw_input()

	tree = [0 for i in range(2*length-1)]

	create_tree(tree, input_string, 0, 0, length - 1)

	no_test_cases = int(raw_input())

	print "Test " + str(count+1) + ":\n"

	for j in range (0, no_test_cases):
		curr_type = int(raw_input())
		if curr_type == 0:

			if tree[0].total_sum == 0 and tree[1].total_sum >= 0 and tree[2].total_sum <= 0:
				print "YES\n"
			else:
				print "NO\n"
		else:
			update_tree(tree, length - 2 + curr_type)

File: spoj/BRCKTS.cpp
#include <iostream>
#include <string>
#include <stdio.h>
#include <math.h>

using namespace std;

struct Node
{
	int openBracketsCount, closedBracketsCount, totalSum, startIndex, endIndex;
};

void createTree(Node tree[], string inputBracketsString, int currentNodeIndex, int startIndex, int endIndex)
{
	if (startIndex > endIndex)
	{
		return;
	}

	//leaf node
	if (startIndex == endIndex)
	{
		Node n;

		if (inputBracketsString.at(startIndex) == '(')
		{
			n.openBracketsCount = 1;
			n.closedBracketsCount = 0;
			n.totalSum = 1;
		}
		else
		{
			n.openBracketsCount = 0;
			n.closedBracketsCount = 1;
			n.totalSum = -1;
		}

		n.startIndex = startIndex;
		n.endIndex = startIndex;
		tree[currentNodeIndex] = n;

		return;
	}

	int mid = (startIndex + endIndex) /2;

	createTree(tree, inputBracketsString, currentNodeIndex*2 + 1, startIndex, mid);
	createTree(tree, inputBracketsString, currentNodeIndex*2 + 2, mid + 1, endIndex);

	Node leftChild =  tree[currentNodeIndex*2 + 1];
	Node rightChild = tree[currentNodeIndex*2 + 2];

	Node currNode;
	currNode.openBracketsCount = leftChild.openBracketsCount + rightChild.openBracketsCount;
	currNode.closedBracketsCount = leftChild.closedBracketsCount + rightChild.closedBracketsCount;
	currNode.totalSum = leftChild.totalSum + rightChild.totalSum;
	currNode.startIndex = leftChild.startIndex;
	currNode.endIndex = rightChild.endIndex;

	tree[currentNodeIndex] = currNode;
}

void updateTree(Node tree[], int updateIndex)
{
	if (updateIndex < 0)
	{
		return;
	}

	Node currentNode = tree[updateIndex];

	if (currentNode.startIndex == currentNode.endIndex)
	{
		if (currentNode.openBracketsCount == 1)
		{
			currentNode.openBracketsCount = 0;
			currentNode.closedBracketsCount = 1;
			currentNode.totalSum = -1;
		}
		else
		{
			currentNode.openBracketsCount = 1;
			currentNode.closedBracketsCount = 0;
			currentNode.totalSum = 1;	
		}
	}
	else
	{
		Node leftChild =  tree[updateIndex*2 + 1];
		Node rightChild = tree[updateIndex*2 + 2];

		currentNode.openBracketsCount = leftChild.openBracketsCount + rightChild.openBracketsCount;
		currentNode.closedBracketsCount = leftChild.closedBracketsCount + rightChild.closedBracketsCount;
		currentNode.totalSum = leftChild.totalSum + rightChild.totalSum;
		currentNode.startIndex = leftChild.startIndex;
		currentNode.endIndex = rightChild.endIndex;
	}

	tree[updateIndex] = currentNode;

	if (updateIndex % 2 == 0)
	{
		updateTree(tree, updateIndex/2 - 1);
	}
	else
	{
		updateTree(tree, floor(updateIndex/2));	
	}
}

int main(int argc, char const *argv[])
{
	#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin); 
	#endif

	for (int i = 0 ; i < 10 ; i++)
	{
		int lengthOfInputString;
		cin >> lengthOfInputString;

		string inputString;

		cin >> inputString;

		Node tree[2*lengthOfInputString];

		cout << "creating tree\n";

		createTree(tree, inputString, 0, 0, lengthOfInputString - 1);

		cout << "Test " << (i+1) << ":\n"; 

		int noProcessing;
		cin >> noProcessing;

		int type;

		for (int j = 0 ; j < noProcessing ; j++)
		{
			cin >> type;
			cout << "type is: " << type << "\n";

			if (type == 0)
			{
				if (tree[0].totalSum == 0 && tree[1].totalSum >= 0 && tree[2].totalSum <= 0)
				{
					cout << "YES\n";
				}
				else
				{
					cout << "NO\n";
				}
			}
			else
			{
				updateTree(tree, lengthOfInputString - 2 + type);
			}
		}
	}
	
	return 0;
}